### DISCUSSION TITLE
Continuous feature의 embedding vector 구하는 방법

### DISCUSSION THREAD

**Discussion by PJHgh** [bcaitech1/p4-dkt-team-ikyo#5] [Ideas]
#### 1. Label Encoder

   sklearn의 LabelEncoder 클래스를 이용하면 continuous feature들도 int 형의 categorical feature로 바꿔줄 수 있다. 이렇게 변환 후 embedding vector를 구하면 될 것 같다.

   ```python
   from sklearn.preprocessing import LabelEncoder

   def encode(df,cols):
       enc =  {}
       for col in cols:
           print(col)
           lbencoder = LabelEncoder()
           lb = lbencoder.fit(df[col].values)
           df[col]=lb.transform(df[col].values)
           enc[col]=lb

       return df, enc

   cat_columns = ['prior_question_had_explanation', 'bundle_id', 'part', 'tags1', 'tags2', 'tags3']

   X[cat_columns].head()
   ```

   [output]
   ![image](https://user-images.githubusercontent.com/51353039/119515757-18ed5480-bdb1-11eb-9f6f-e89db3f3b710.png)

   ```python
   encode_X,enc_dict = encode(X, cat_columns)
   encode_X[cat_columns].head()
   ```

   [output]
   ![image](https://user-images.githubusercontent.com/51353039/119515855-2e627e80-bdb1-11eb-8b60-731c59944a63.png)

   [Label Encoder](https://pinkwink.kr/1247#:~:text=%EB%9D%BC%EB%B2%A8%EC%9D%B8%EC%BD%94%EB%94%A9%EC%9D%B4%EB%9D%BC%EB%8A%94%20%EA%B2%83%EC%9D%80%20%EC%9D%B4%EB%A6%84,%EC%9D%98%20LabelEncoder%EB%A5%BC%20import%20%ED%95%A9%EB%8B%88%EB%8B%A4).

#### 2. Dense Layer

   continuous feature를 그대로 사용하는 방법도 있다. 하지만 사용하기 전 미리 scaling을 해줘야 하는 것 같다. scaling도 sklearn에서 지원해주는 여러 Scaler 중 하나를 사용하면 될 것 같고, 마지막으로 batch normalization을 통해 전체 continuous feature의 평균을 0, 분산을 1로 만드는 normalization을 해준 뒤 사용하면 되는 것 같다.

   - Scaler

     ```python
     cont_columns = ['prior_question_elapsed_time', "content_mean", "content_std", "content_sum", "content_skew", "user_mean", 'user_sum',  'user_std', 'user_skew']

     X[cont_columns].head()
     ```

     [output]
     ![image](https://user-images.githubusercontent.com/51353039/119515928-3f12f480-bdb1-11eb-8cb4-d488fc81d9c4.png)

     ```python
     scale_dict={}
     fix_missing={}
     for col in cont_columns:
         scaler = RobustScaler()
         scale_dict[col]=scaler.fit(X[col].values.reshape(-1,1))
         X[col] = scale_dict[col].transform(X[col].values.reshape(-1,1))
         fix_missing[col] = X[col].mode().values[0]
     ```

     [output]
     ![image](https://user-images.githubusercontent.com/51353039/119515967-49cd8980-bdb1-11eb-9b67-3dd3791109d5.png)

     [SKLearn의 Scaler](https://mkjjo.github.io/python/2019/01/10/scaler.html#:~:text=4.-,RobustScaler,%EC%9D%84%20%ED%99%95%EC%9D%B8%20%ED%95%A0%20%EC%88%98%20%EC%9E%88%EB%8B%A4).
   - Batch Norm

     ```python
     class RidModel(nn.Module):
         def __init__(self,emb_dims,no_of_cont):
             super(RidModel, self).__init__()

             self.emb = nn.ModuleList([nn.Embedding(x,y) for x,y in emb_dims])

             no_of_embs = sum([y for x, y in emb_dims])
             self.no_of_embs = no_of_embs
             self.no_of_cont = no_of_cont

             self.batch_norm1 = nn.BatchNorm1d(self.no_of_cont)
             self.dropout1 = nn.Dropout(0.2)
             self.dense1 = nn.utils.weight_norm(nn.Linear(no_of_cont, 128))

         ...
         def forward(self, cont,cat):

             ## cat data part
             x_cat = [emb_layer(cat[:,i]) for i,emb_layer in enumerate(self.emb)]
             x_cat = torch.cat(x_cat,1)
             x_cat = self.dropout1(x_cat)

             ##cont data
             x = self.batch_norm1(cont)
             x = self.dropout1(x)
             x = F.relu(self.dense1(x))
         ...
     ```
#### 3. Drop out

   위 코드를 보며 embedding vector를 구하는 방법을 알아보다가 좋은 모델 학습 방법을 알게 되었다. 각 feature의 embedding vector를 구한 다음 위 모델 코드와 같이 drop out을 사용하는 방법이다. 이렇게 사용하게 되면 이미지에 약간의 노이즈를 주어 학습하는 방법이나 nlp에서 random하게 masking시키는 방법과 동일한 효과를 볼 수 있을 것 같다.

[퀸수지님 방명록]
![image](https://user-images.githubusercontent.com/51353039/119516006-53ef8800-bdb1-11eb-9b2c-b9ddfbbd9838.png)

[reference]
[Pytorch + entity Embedding🔥🔥](https://www.kaggle.com/shahules/pytorch-entity-embedding#Model)

**Comment by JaeheeRyu** [comment]
킹종헌.. 그는 대체.. 공유해주셔서 감사합니다!!
+) 방명록은 @ohsuz 님이 쓰셨고, 저는 ㅋㅋㅋ만 썼답니다!

**Reply by PJHgh** [reply]
앗....수정하겠습니다ㅎㅎ
